[Form Control Refresh] `accent-color` should preserve control legibility

lilyspiniolas opened the pull request:
#### 4f0c0eb24e9b1b2d4d6e69d6f2b6b60a8420a0e8
<pre>
[Form Control Refresh] `accent-color` should preserve control legibility
<a href="https://bugs.webkit.org/show_bug.cgi?id=244233">https://bugs.webkit.org/show_bug.cgi?id=244233</a>
<a href="https://rdar.apple.com/99018889">rdar://99018889</a>

Reviewed by Aditya Keerthi.

When form control refresh is enabled, controls with tint color now preserve
legibility regardless of any set `accent-color`. To ensure that all controls can display
the same accent-colors, while also taking into account the fact that some controls
may not be able to use certain accent-colors without extensive changes to other parts
of the contol, the visual accent-color is bounded based on whether light mode or dark
mode is being used. In light mode, accent-colors with a luminance greater than 0.5
will have their luminance lowered to 0.5. In dark mode, accent-colors with a luminance
less than or equal to 0.5 will have their luminance increased to 0.5. For both
operations, the color&apos;s hue is preserved.

Additionally, checkboxes and radio buttons will now use a dark indicator in
cases where the accent-color luminance is greater than 0.5.

The threshold of 0.5 luminance was chosen to match AppKit&apos;s threshold for switching
indicator colors for checkboxes and radio buttons.

`setInsideDisabledSubmitButton` is now `setInsideSubmitButton` and is applied to
all submit buttons, regardless of enabled/disabled state in order to properly set
the text color based on enablement, the window active state, and the controls tint
color in order to preserve contrast.

Affected controls:
* &lt;input type=range&gt; (sliders)
* &lt;input type=checkbox&gt;
* &lt;input type=radio&gt;
* &lt;input type=text list=somedatalist&gt; (list buttons)
* &lt;input type=&quot;submit&quot;&gt;
* &lt;progress&gt;

* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-checkbox-radio-do-not-disappear-dark-mode-expected-mismatch.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-checkbox-radio-do-not-disappear-dark-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-checkbox-radio-do-not-disappear-light-mode-expected-mismatch.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-checkbox-radio-do-not-disappear-light-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-does-not-affect-inactive-window-dark-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-does-not-affect-inactive-window-dark-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-does-not-affect-inactive-window-light-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-does-not-affect-inactive-window-light-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-submit-does-not-disappear-expected-mismatch.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-submit-does-not-disappear.html: Added.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::submitButtonTextColor const):
(WebCore::RenderTheme::disabledSubmitButtonTextColor const): Deleted.
* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextPaintStyle):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::foregroundColorForBackgroundColor):
(WebCore::colorWithTargetLuminance):
(WebCore::checkboxRadioIndicatorColorForVectorBasedControls):
(WebCore::adjustCheckboxRadioBackgroundColorDisabledState):
(WebCore::RenderThemeCocoa::checkboxRadioBackgroundColorForVectorBasedControls const):
(WebCore::RenderThemeCocoa::paintCheckboxForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintRadioForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintButtonForVectorBasedControls):
(WebCore::RenderThemeCocoa::adjustButtonStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::controlTintColorWithContrast const):
(WebCore::RenderThemeCocoa::paintListButtonForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintProgressBarForVectorBasedControls):
(WebCore::paintSliderTicksForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintSliderTrackForVectorBasedControls):
(WebCore::RenderThemeCocoa::submitButtonTextColor const):
(WebCore::RenderThemeCocoa::disabledSubmitButtonTextColor const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresRepaint):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::insideSubmitButton const):
(WebCore::RenderStyle::insideDisabledSubmitButton const): Deleted.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setInsideSubmitButton):
(WebCore::RenderStyle::setInsideDisabledSubmitButton): Deleted.
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:

Canonical link: <a href="https://commits.webkit.org/298318@main">https://commits.webkit.org/298318@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a8e48e24c5eedf197daa1d36ae30de5ba9438a1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115149 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34857 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25346 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121260 "Failed to checkout and rebase branch from PR 48944") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/65774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ea0cabec-d8a2-47f3-a163-298d490a8e70) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117038 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35509 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43423 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/121260 "Failed to checkout and rebase branch from PR 48944") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/65774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/eec31269-1d26-4507-a071-14eed1e99ab2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118097 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/28307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/103378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/121260 "Failed to checkout and rebase branch from PR 48944") | | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/27475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/21498 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64915 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/97691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/21615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/124446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42110 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/31502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/124446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42481 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/99568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/124446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/41287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/19134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18417 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41988 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47529 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/41518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/44840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/43250 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/RenderTheme.h:
```diff
@@ -263,7 +263,7 @@ class RenderTheme {
     virtual void adjustTextControlInnerTextStyle(RenderStyle&, const RenderStyle&, const Element*) const { }
 
-    virtual Color disabledSubmitButtonTextColor() const { return Color::black; }
+    virtual Color submitButtonTextColor(const RenderObject&) const { return Color::black; }
```
Is this a distinct behavior change? Not sure how it relates to this PR. If it does, could use some explanation in the commit message.

lilyspiniolas replied:
yes, I'll update the message

pxlcoder commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -171,6 +171,29 @@ static void drawHighContrastOutline(GraphicsContext& context, Path path, OptionS
+    return backgroundColor.luminance() <= controlTintLuminanceThreshold ? WebCore::Color::white : WebCore::Color::black;
+}
+
+static Color forceLuminance(Color color, float targetLuminance)
```
Rename to `colorWithTargetLuminance`?

pxlcoder commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -171,6 +171,29 @@ static void drawHighContrastOutline(GraphicsContext& context, Path path, OptionS
 }
 #endif
 
+static constexpr auto controlTintLuminanceThreshold = 0.5f;
```
The reason behind using this specific value should be explained in the commit message.

pxlcoder commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -2072,6 +2113,11 @@ static void paintTextAreaOrTextFieldInnerShadow(GraphicsContext& context, const
 }
 #endif
 
+static Color textAreaOrTextFieldBackgroundColor(const OptionSet<StyleColorOptions> styleColorOptions)
```
I only see one caller, is this necessary?

nt1m reviewed (commented):
The tests have inconsistent indentation. I typically indent everything except the `html` tag

lilyspiniolas committed:
[Form Control Refresh] `accent-color` should preserve control legibility
https://bugs.webkit.org/show_bug.cgi?id=244233
rdar://99018889

Reviewed by Aditya Keerthi.

When form control refresh is enabled, controls with tint color now preserve
legibility regardless of any set `accent-color`. To ensure that all controls can display
the same accent-colors, while also taking into account the fact that some controls
may not be able to use certain accent-colors without extensive changes to other parts
of the contol, the visual accent-color is bounded based on whether light mode or dark
mode is being used. In light mode, accent-colors with a luminance greater than 0.5
will have their luminance lowered to 0.5. In dark mode, accent-colors with a luminance
less than or equal to 0.5 will have their luminance increased to 0.5. For both
operations, the color's hue is preserved.

Additionally, checkboxes and radio buttons will now use a dark indicator in
cases where the accent-color luminance is greater than 0.5.

The threshold of 0.5 luminance was chosen to match AppKit's threshold for switching
indicator colors for checkboxes and radio buttons.

`setInsideDisabledSubmitButton` is now `setInsideSubmitButton` and is applied to
all submit buttons, regardless of enabled/disabled state in order to properly set
the text color based on enablement, the window active state, and the controls tint
color in order to preserve contrast.

Affected controls:
* <input type=range> (sliders)
* <input type=checkbox>
* <input type=radio>
* <input type=text list=somedatalist> (list buttons)
* <input type="submit">
* <progress>

* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-checkbox-radio-do-not-disappear-dark-mode-expected-mismatch.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-checkbox-radio-do-not-disappear-dark-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-checkbox-radio-do-not-disappear-light-mode-expected-mismatch.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-checkbox-radio-do-not-disappear-light-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-does-not-affect-inactive-window-dark-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-does-not-affect-inactive-window-dark-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-does-not-affect-inactive-window-light-mode-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-does-not-affect-inactive-window-light-mode.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-submit-does-not-disappear-expected-mismatch.html: Added.
* LayoutTests/fast/forms/form-control-refresh/accent-color-contrast-submit-does-not-disappear.html: Added.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::submitButtonTextColor const):
(WebCore::RenderTheme::disabledSubmitButtonTextColor const): Deleted.
* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextPaintStyle):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::foregroundColorForBackgroundColor):
(WebCore::colorWithTargetLuminance):
(WebCore::checkboxRadioIndicatorColorForVectorBasedControls):
(WebCore::adjustCheckboxRadioBackgroundColorDisabledState):
(WebCore::RenderThemeCocoa::checkboxRadioBackgroundColorForVectorBasedControls const):
(WebCore::RenderThemeCocoa::paintCheckboxForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintRadioForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintButtonForVectorBasedControls):
(WebCore::RenderThemeCocoa::adjustButtonStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::controlTintColorWithContrast const):
(WebCore::RenderThemeCocoa::paintListButtonForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintProgressBarForVectorBasedControls):
(WebCore::paintSliderTicksForVectorBasedControls):
(WebCore::RenderThemeCocoa::paintSliderTrackForVectorBasedControls):
(WebCore::RenderThemeCocoa::submitButtonTextColor const):
(WebCore::RenderThemeCocoa::disabledSubmitButtonTextColor const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresRepaint):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::insideSubmitButton const):
(WebCore::RenderStyle::insideDisabledSubmitButton const): Deleted.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setInsideSubmitButton):
(WebCore::RenderStyle::setInsideDisabledSubmitButton): Deleted.
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:

Canonical link: https://commits.webkit.org/298318@main
